### PR TITLE
Adding channel LCO

### PR DESCRIPTION
**Pull request by hkaiser**
This PR adds two new LCOs: local::channel and channel (one for local only operation, the other is a component). Both expose a similar interface which is aligned with the channel object known from the Go language.

This is meant to be a basis for discussions aiming at finding the best possible API for such a facility, so please comment.

### CODE REVIEW

File: docs/manual/lcos.qbk [hunk 1]
```diff
@@ -20,11 +20,73 @@ special different ways adapted to HPX.
...
+        hpx::lcos::local::receive_channel<int> c,
+        hpx::lcos::local::send_channel<> done)
+    {
+        cout << c.get();        // prints 42
+        done.set();             // signal back
+    }
+
+    {
+        hpx::lcos::local::channel<int> c;
+        hpx::lcos::local::channel<> done;
```
Comment (sithhell) [inline docs/manual/lcos.qbk]: why do the types not match in the function signature and here?

File: examples/quickstart/channel.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    pongs.set(msg);
+}
+
+void pingpong()
+{
+    hpx::lcos::local::channel<std::string> pings;
+    hpx::lcos::local::channel<std::string> pongs;
+
+    ping(pings, "passed message");
+    pong(pings, pongs);
```
Comment (sithhell) [inline examples/quickstart/channel.cpp]: where is the conversion done? what are the rules?

File: hpx/lcos/channel.hpp [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+
+    public:
+        channel()
+        {}
+
+        // create a new instance of a channel component
+        channel(naming::id_type const& loc)
+          : base_type(hpx::new_<lcos::server::channel<T> >(loc))
+        {}
+        channel(lcos::future<naming::id_type> f)
```
Comment (sithhell) [inline hpx/lcos/channel.hpp]: Does this really need to be public? I would find it nice if we could avoid things like these.

File: hpx/lcos/channel.hpp [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+
+        HPX_CONSTEXPR_OR_CONST
+        static std::size_t default_generation = std::size_t(-1);
+
+    public:
+        channel()
+        {}
+
+        // create a new instance of a channel component
+        channel(naming::id_type const& loc)
```
Comment (sithhell) [inline hpx/lcos/channel.hpp]: shouldn't this be explicit?

File: hpx/lcos/channel.hpp [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+        T get(std::size_t generation = default_generation,
+            hpx::error_code& ec = hpx::throws) const
+        {
+            return get_async(generation).get(ec);
+        }
+        T get(hpx::error_code& ec,
+            std::size_t generation = default_generation) const
+        {
+            return get_async(generation).get(ec);
+        }
```
Comment (sithhell) [inline hpx/lcos/channel.hpp]: I would rather have `get` to return the future, and have an additional `get_sync`, if necessary.

File: hpx/lcos/channel.hpp [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+
+        ///////////////////////////////////////////////////////////////////////
+        template <typename U, typename U2 = T, typename Enable =
+            typename std::enable_if<!std::is_void<U2>::value>::type>
+        hpx::future<void>
+        set_async(U val, std::size_t generation = default_generation)
```
Comment (sithhell) [inline hpx/lcos/channel.hpp]: why don't you use perfect forwarding for `val`?

File: hpx/lcos/channel.hpp [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+        {
+            typedef typename lcos::server::channel<T>::set_generation_action
+                action_type;
+            return hpx::async(action_type(), this->get_id(), std::move(val),
```
Comment (sithhell) [inline hpx/lcos/channel.hpp]: Isn't set a fire and forget operation?

File: hpx/lcos/channel.hpp [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+
+        channel_iterator(Channel const& c)
+          : channel_(&c), data_(get_checked())
+        {}
+
+    private:
+        bool get_checked()
+        {
+            hpx::future<void> f = channel_->get_async();
+            f.wait();
```
Comment (sithhell) [inline hpx/lcos/channel.hpp]: Can't we 'futurize' the iterator?

File: hpx/lcos/server/queue.hpp [hunk 9]
```diff
@@ -34,14 +35,14 @@ namespace hpx { namespace lcos { namespace server
     template <typename ValueType, typename RemoteType>
     class queue
       : public lcos::base_lco_with_value<ValueType, RemoteType>
-      , public components::component_base<queue<ValueType, RemoteType> >
+      , public components::managed_component_base<queue<ValueType, RemoteType> >
```
Comment (sithhell) [inline hpx/lcos/server/queue.hpp]: What's the rationale for this change?

File: docs/manual/lcos.qbk (hunk 1 above)
Comment (hkaiser) [inline docs/manual/lcos.qbk]: They could match. `receive_channel` and `send_channel` are constructable from `channel`, but they limit its use to one end of the communication. Thus this would work as well:

    void do_something(  
        hpx::lcos::local::channel<int> c,  
        hpx::lcos::local::channel<> done)  
    { ... }

File: examples/quickstart/channel.cpp (hunk 2 above)
Comment (hkaiser) [inline examples/quickstart/channel.cpp]: As said, `receive_channel` and `send_channel` are implicitly constructable from `channel`.

File: hpx/lcos/channel.hpp (hunk 3 above)
Comment (hkaiser) [inline hpx/lcos/channel.hpp]: This is necessary for the following code to work (currently):

    hpx::lcos::channel c;
    c = hpx:new_<hpx::lcos::server::channel>(hpx::find_here());

Thus it's enforced by the current creation APIs we have. If we changed those rules this public constructor could be avoided.

File: hpx/lcos/channel.hpp (hunk 4 above)
Comment (hkaiser) [inline hpx/lcos/channel.hpp]: Fair point.

File: hpx/lcos/channel.hpp (hunk 5 above)
Comment (hkaiser) [inline hpx/lcos/channel.hpp]: Fine by me, either way is ok.

File: hpx/lcos/channel.hpp (hunk 6 above)
Comment (hkaiser) [inline hpx/lcos/channel.hpp]: The underlying action expects an rvalue, currently.

File: hpx/lcos/channel.hpp (hunk 8 above)
Comment (hkaiser) [inline hpx/lcos/channel.hpp]: Nod, I was thinking about it, but without `await` in the language I wouldn't see a way of making this work.

File: hpx/lcos/server/queue.hpp (hunk 9 above)
Comment (hkaiser) [inline hpx/lcos/server/queue.hpp]: All components which are LCOs (which are derived from base_lco) have to be managed components.

File: hpx/lcos/channel.hpp (hunk 3 above)
Comment (sithhell) [inline hpx/lcos/channel.hpp]: On Montag, 25. Juli 2016 11:13:20 CEST Hartmut Kaiser wrote:
> > +        typedef components::client_base<
> > +                channel<T>, lcos::server::channel<T>
> > +            > base_type;
> > +
> > +        HPX_CONSTEXPR_OR_CONST
> > +        static std::size_t default_generation = std::size_t(-1);
> > +
> > +    public:
> > +        channel()
> > +        {}
> > +
> > +        // create a new instance of a channel component
> > +        channel(naming::id_type const& loc)
> > +          : base_type(hpx::new_<lcos::server::channel<T> >(loc))
> > +        {}
> > +        channel(lcos::future<naming::id_type> f)
> 
> This is necessary for the following code to work (currently):
> 
>     hpx::lcos::channel c;
>     c = hpx:new_<hpx::lcos::server::channel>(hpx::find_here());
> 
> Thus it's enforced by the current creation APIs we have. If we changed those
> rules this public constructor could be avoided.

Well, this form isn't really needed as the first constructor does exactly that, 
no? I wouldn't think that our current creation APIs enforce it. It can be 
completely hidden, as your first ctor has shown.

Comment (hkaiser) [inline hpx/lcos/channel.hpp]: > Well, this form isn't really needed as the first constructor does exactly that, no? I wouldn't think that our current creation APIs enforce it. It can be completely hidden, as your first ctor has shown.

Certainly you can hide it for the use case where you directly construct the component at client construction time. The question however is whether this is the only way one might want to create the component.

I'd be fine for now with removing this constructor if we're sure that this wouldn't limit the overall usability of the facility.

File: hpx/lcos/channel.hpp (hunk 5 above)
Comment (hkaiser) [inline hpx/lcos/channel.hpp]: I'm still not sure how to name functions. Essentially, we have `get`, `set`, and `close`. The `get` function could (and probably should) be asynchronous by default for the local and the remote version of channel. 

However, what about `set`? I don't think it should have fire&forget semantics as there wouldn't be a way of detecting certain errors. Also, for the local version there it does not really make sense way of making it asynchronous (at least not for an 'unlimited' channel). The remote version can be made asynchronous as it calls `async` underneath.

We have a similar situation with `close`. What should be the default semantics? sync? async? fire&forget?

Would it be ok to have different semantics for `get`, `set`, and `close`?

Comment (hkaiser) [inline hpx/lcos/channel.hpp]: I have made `get` asynchronous by default but have left `set` and `close` unchanged. Would that be what you'd expect?

File: docs/manual/lcos.qbk (hunk 1 above)
Comment (sithhell) [inline docs/manual/lcos.qbk]: I guess what I wanted to say is that this information (implicit conversion and purpose) is missing from the documentation.

File: hpx/lcos/channel.hpp (hunk 6 above)
Comment (sithhell) [inline hpx/lcos/channel.hpp]: Even if the action expects an rvalue, we still create a packaged action first, which will benefit from perfect forwarding. The execution of the action is a one shot operation anyway.

File: docs/manual/lcos.qbk (hunk 1 above)
Comment (hkaiser) [inline docs/manual/lcos.qbk]: The docs are currently sketchy at best anyways ;-)

File: hpx/lcos/channel.hpp (hunk 5 above)
Comment (sithhell) [inline hpx/lcos/channel.hpp]: Having `get` asynchronously is what I'd expect, yes. 
For the rest, I am not sure either... The fire&forget suggestions for `set` was mainly due to performance considerations. What errors could occur? Do we care on the `set` side of things? An alternative approach to member functions might be some form of free functions maybe, or having different policies: ser -> synchronous, par -> fire&forget, par(task) -> returning a future?. I think different people will have different expectations from this utility.

File: hpx/lcos/channel.hpp (hunk 3 above)
Comment (sithhell) [inline hpx/lcos/channel.hpp]: What use cases do you have in mind? Connecting to an already created component is handled by the client interface. The only thing I can think of right now is colocated creation.
The biggest problem I have with this (but that is mostly a general concern) is that both ctors look almost the same, it is not clear from it's signature where the differences are. I'd rather not have the additional future<id_type> ctor in the public interface.

File: hpx/lcos/channel.hpp (hunk 5 above)
Comment (hkaiser) [inline hpx/lcos/channel.hpp]: I was thinking of re-using the execution policies for this as well, it's a compelling thought. It would be a lot of work, though, as we have a lot of spots where this could be applied...

File: hpx/lcos/channel.hpp (hunk 3 above)
Comment (hkaiser) [inline hpx/lcos/channel.hpp]: I think I've already removed the constructor.

File: hpx/lcos/channel.hpp (hunk 8 above)
Comment (sithhell) [inline hpx/lcos/channel.hpp]: True, without `co_await`, this is hard to implement. Why not have `future<T>` as value type directly? This won't prohibit the use of `co_await` in future code, and it's the user's responsibility to deal with the future: block immediately, or create some kind of awkward continuation based looping infrastructure. I am wondering if we can't offer a high level solution to loop over future values though....

File: hpx/lcos/channel.hpp [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+    private:
+        std::pair<T, bool> get_checked() const
+        {
+            hpx::future<T> f = channel_->get();
+            f.wait();
+            if (!f.has_exception())
+            {
+                return std::make_pair(f.get(), true);
+            }
+            return std::make_pair(T(), false);
```
Comment (sithhell) [inline hpx/lcos/channel.hpp]: This is swalloing exceptions, isn't it?

Comment (hkaiser) [inline hpx/lcos/channel.hpp]: Yes, this is used to flag the end of a range, so it ignores the exceptions. This is yet another argument in favor of making the iterators lazy.

**Comment by hkaiser** [comment]
@sithhell How should we proceed here? Is this ok to be merged?

**Comment by sithhell** [comment]
@hkaiser I am not sure, how should we resolve the remaining open points?

**Comment by hkaiser** [comment]
@sithhell The only thing I can try to work on is to make the iterators lazy. Everything else requires decisions beyond the scope of this PR.

**Comment by hkaiser** [comment]
Some more investigation has shown that implementing asyncronous iterators requires building a completely new infrastructure which is way beyond the scope of this PR. Thus I'd like to go ahead with this as is now. If the current iterator design is too controversial I'd be happy to remove it alltogether for now.

**Comment by hkaiser** [comment]
[07:05]	heller: regarding channel ... should we just ditch the iterator? do you have a specific use case?
[07:05]	heller: if you want to keep it ... let's keep it synchronous
[07:06]	hkaiser: heller: I want to work on a async iterator framework, but that requires more time
[07:06]	heller: *nod*
[07:06]	hkaiser: I have no problems ditching the iterator for now
[07:07]	heller: I am in favor for having the async iterator framework first
[07:07]	heller: nevertheless, I have no problem keeping it synchronous for now
[07:07]	hkaiser: ok
[07:07]	hkaiser: I'll leave it in

**Comment by hkaiser** [comment]
Once #2276 has been merged we can change the interfaces for the `set`/`set_sync`/`set_apply` functions, etc.

**Comment by hkaiser** [comment]
This PR is ready to be merged now. The channel API has been adapted to the new launch policies.

File: docs/manual/lcos.qbk (hunk 1 above)
Comment (sithhell) [inline docs/manual/lcos.qbk]: Right, so why not make them good from the start ;)?

File: examples/quickstart/channel.cpp [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+void calculate_sum()
+{
+    std::vector<int> s = { 7, 2, 8, -9, 4, 0 };
+    hpx::lcos::local::channel<int> c;
+
+    hpx::apply(&sum, std::vector<int>(s.begin(), s.begin() + s.size()/2), c);
+    hpx::apply(&sum, std::vector<int>(s.begin() + s.size()/2, s.end()), c);
+
+    int x = c.get_sync();    // receive from c
+    int y = c.get_sync();
```
Comment (sithhell) [inline examples/quickstart/channel.cpp]: The two lines above don't look right. Why don't you use launch policies there?

File: examples/quickstart/channel.cpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    std::string const& msg)
+{
+    pings.set(msg);
+}
+
+void pong(
+    hpx::lcos::local::receive_channel<std::string> pings,
+    hpx::lcos::local::send_channel<std::string> pongs)
+{
+    std::string msg = pings.get_sync();
```
Comment (sithhell) [inline examples/quickstart/channel.cpp]: Same as above.

File: examples/quickstart/channel.cpp [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+{
+    hpx::lcos::local::channel<int> jobs;
+    hpx::lcos::local::channel<> done;
+
+    hpx::apply(
+        [jobs, done]() mutable
+        {
+            while(true)
+            {
+                auto next = jobs.get_checked();
```
Comment (sithhell) [inline examples/quickstart/channel.cpp]: could you add a comment on how `get_checked()` is different from just `get()`? Why do we need both? When to use what?

File: hpx/lcos/channel.hpp [hunk 14]
```diff
@@ -0,0 +1,481 @@
...
+        }
+        void close(launch::sync_policy)
+        {
+            typedef typename lcos::server::channel<T>::close_action action_type;
+            action_type()(this->get_id());
+        }
+        void close()
+        {
+            close(launch::sync);
+        }
```
Comment (sithhell) [inline hpx/lcos/channel.hpp]: would be nice to have an apply_policy overload here as well.

File: hpx/lcos/local/channel.hpp [hunk 15]
```diff
@@ -0,0 +1,672 @@
...
+
+            HPX_NON_COPYABLE(unlimited_channel);
+
+        public:
+            unlimited_channel()
+              : closed_(false)
+            {}
+
+        protected:
+            hpx::future<T> get(std::size_t generation, bool blocking)
```
Comment (sithhell) [inline hpx/lcos/local/channel.hpp]: blocking isn't used in that function. What are the semantics of a function returning a future that blocks?

File: examples/quickstart/channel.cpp (hunk 11 above)
Comment (hkaiser) [inline examples/quickstart/channel.cpp]: Wyy should I?

Comment (hkaiser) [inline examples/quickstart/channel.cpp]: Sorry, I forgot to change the example.

File: hpx/lcos/local/channel.hpp (hunk 15 above)
Comment (hkaiser) [inline hpx/lcos/local/channel.hpp]: This is an internal argument. It informs the implementation that the caller is blocking which changes the error handling inside the implementation. We might want to revisit this, it's not pretty.

File: hpx/lcos/channel.hpp (hunk 14 above)
Comment (hkaiser) [inline hpx/lcos/channel.hpp]: Ok, can do.

Comment (hkaiser) [inline hpx/lcos/channel.hpp]: This has been added

File: hpx/lcos/local/channel.hpp (hunk 15 above)
Comment (hkaiser) [inline hpx/lcos/local/channel.hpp]: An alternative for this Boolean argument would be to expose two different functions from the implementation, but those two functions would be nearly identical.
